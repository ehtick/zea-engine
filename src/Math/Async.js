import { Signal } from '../Math/Signal';

class Async {
    constructor() {
        this.__asyncCount = 0;
        this.ready = new Signal(true);

        this.incAsyncCount = function(count=1) {
            this.__asyncCount+=count;
        }.bind(this);

        this.decAsyncCount = function() {
            if (this.__asyncCount > 0){
                this.__asyncCount--;
                if (this.__asyncCount == 0) {
                    this.__asyncsCompleted();
                }
            }
        }.bind(this);

        this.__asyncsCompleted = function(){
            this.ready.emit();
        }.bind(this)
    }

    get count(){
        return this.__asyncCount;
    }

};

export {
    Async
};
// export default Async;

