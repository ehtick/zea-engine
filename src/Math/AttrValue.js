// eslint-disable-next-line camelcase
import { JSON_stringify_fixedPrecision } from './Common.js'

/**
 * Base class for values that can be stored in vertex attributes.
 * <br>
 * **Note:** These values use {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Float32Array|Float32Array} values to store their data.
 */
class AttrValue {
  /**
   * Verifies that all the numeric values inside the `this.__data` variable are a valid number.
   * Returns `false` If at least one of the values is either {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Referencia/Objetos_globales/Infinity|Infinity} or
   * {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Referencia/Objetos_globales/NaN|NaN}.
   *
   * @return {boolean} - Returns the result as a boolean.
   */
  isValid() {
    for (const v of this.__data) {
      if (v == Infinity || isNaN(v)) return false
    }

    return true
  }

  /**
   * Creates a new value to wrap memory in an existing buffer.
   *
   * @param {ArrayBuffer} buffer - the buffer value.
   * @param {number} offset - the offset value.
   *
   * @todo Implement this function
   */
  static createFromFloat32Buffer(buffer, offset) {
    throw new Error(
      'Not yet implemented for this type:' + this.constructor.name
    )
  }

  /**
   * Returns the number of Float32 elements used by this type. Used to calculate storage requirements for large arrays of this type.
   *
   * @todo Implement this function
   */
  static numElements() {
    throw new Error(
      'Not yet implemented for this type:' + this.constructor.name
    )
  }

  /**
   * Returns current Vec2 data as array. Often used to pass types to the GPU.
   *
   * @return {array} - Returns the result as an array.
   */
  asArray() {
    return this.__data
  }

  /**
   * Calls `toJSON` method and stringifies it.
   *
   * @return {string} - The return value.
   */
  toString() {
    // eslint-disable-next-line new-cap
    return JSON_stringify_fixedPrecision(this.toJSON())
  }

  /**
   * Represents the state of your class as a JSON Object.
   *
   * @todo Implement this function in the derived class.
   */
  toJSON() {
    throw new Error(
      'Not yet implemented for this type:' + this.constructor.name
    )
  }
}

export { AttrValue }
export default AttrValue
