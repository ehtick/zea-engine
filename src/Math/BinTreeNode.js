import Vec2 from './Vec2.js';

class BinTreeRectBorder {
    constructor(value, movable = false) {
        this.value = value;
        this.movable = movable;
    }
};

class BinTreeRect {
    constructor(pos, right, top) {
        this.pos = pos;
        this.right = right;
        this.top = top;
    }

    get size() {
        return new Vec2(this.right.value - this.pos.x, this.top.value - this.pos.y);
    }
};

class BinTreeNode {
    constructor(rect, rootNode = false) {
        this.rect = rect;
        this.child0 = undefined;
        this.child1 = undefined;
        this.occupied = false;
        this.rootNode = rootNode;
        if (this.rootNode) {
            this.rect.right.ownerNode = this;
            this.rect.top.ownerNode = this;
        }
    }

    insert(rectSize, closestFit = undefined) {
        if (this.child0 != undefined && this.child1 != undefined) {
            // try inserting into first child
            let res = this.child0.insert(rectSize, closestFit);
            if (res != undefined)
                return res

            // no room, insert into second.
            return this.child1.insert(rectSize, closestFit);
        } else {

            // if there's already a rectSize here, return...
            if (this.occupied) {
                return undefined;
            }
            let delta = rectSize.subtract(this.rect.size);

            // (if we're too small, return)

            // Using a custom epsilon after packing was failing on simple cases. 
            // See TestBinTree2
            let eps = 0.00001;
            if (delta.x > eps || delta.y > eps) {
                // If we can fit the cluster into the node by moving only one of the sides.
                // Check for movable sides, and if so return this as a possibility of expanding the canvas.
                if (closestFit &&
                    ((this.rect.right.movable && delta.x > 0 && delta.y <= 0) ||
                     (this.rect.top.movable && delta.y > 0 && delta.x <= 0) ||
                     (this.rect.right.movable && delta.x > 0 && this.rect.top.movable && delta.y > 0 ))) {
                    
                    // Generate a cost metric to find the node that requires 
                    // the smallest ajustment of the borders.
                    let cost = 0;
                    if (this.rect.right.movable && delta.x > 0)
                        cost += delta.x;
                    if (this.rect.top.movable && delta.y > 0)
                        cost += delta.y;

                    if (!closestFit.node || cost < closestFit.cost) {
                        closestFit.node = this;
                        closestFit.delta = delta;
                        closestFit.cost = cost;
                    }
                }
                return undefined;
            }

            // if we're just right, accept..
            if (Math.abs(delta.x) < eps && Math.abs(delta.y) < eps) {
                if(this.rect.top.movable){
                    this.sliceHorizontal(this.rect.top.value);
                    this.child0.setOccupied();
                    return this.child0;
                }
                if(this.rect.right.movable){
                    this.sliceVertical(this.rect.right.value);
                    this.child0.setOccupied();
                    return this.child0;
                }
                this.setOccupied();
                return this;
            }

            // otherwise, gotta split this node and create some kids
            if (delta.x < delta.y)
                this.sliceVertical(this.rect.pos.x + rectSize.x);
            else
                this.sliceHorizontal(this.rect.pos.y + rectSize.y);

            return this.child0.insert(rectSize, closestFit);
        }
    }

    setOccupied() {
        this.occupied = true;
    }

    sliceVertical(xDivisor) {
        let child0rect = new BinTreeRect(this.rect.pos, new BinTreeRectBorder(xDivisor), this.rect.top);
        let child1rect = new BinTreeRect(new Vec2(xDivisor, this.rect.pos.y), this.rect.right, this.rect.top);
        this.child0 = new BinTreeNode(child0rect);
        this.child1 = new BinTreeNode(child1rect);
    }

    sliceHorizontal(yDivisor) {
        let child0rect = new BinTreeRect(this.rect.pos, this.rect.right, new BinTreeRectBorder(yDivisor));
        let child1rect = new BinTreeRect(new Vec2(this.rect.pos.x, yDivisor), this.rect.right, this.rect.top);
        this.child0 = new BinTreeNode(child0rect);
        this.child1 = new BinTreeNode(child1rect);
    }

    resizeAndInsert(rectSize, delta) {
        if (delta.x > 0 && this.rect.right.movable && delta.y > 0 && this.rect.top.movable) {
            this.rect.right.value += delta.x;
            this.rect.top.value += delta.y;
            this.sliceVertical(this.rect.right.value);
            this.child0.sliceHorizontal(this.rect.top.value);
            this.child0.child0.setOccupied();
            return this.child0.child0;
        }
        else if (delta.x > 0 && this.rect.right.movable){
            this.rect.right.value += delta.x;
            this.sliceVertical(this.rect.right.value);
            let res = this.child0.insert(rectSize);
            if (res == undefined)
                console.log("We should fit now~!!");
            return res;
        }
        else if (delta.y > 0 && this.rect.top.movable) {
            this.rect.top.value += delta.y;
            this.sliceHorizontal(this.rect.top.value);
            let res = this.child0.insert(rectSize);
            if (res == undefined)
                console.log("We should fit now~!!");
            return res;
        }
        else {
            throw ("Why are we here!!");
        }
    }
};

export {
    BinTreeRect,
    BinTreeRectBorder,
    BinTreeNode
};

export default BinTreeNode;
