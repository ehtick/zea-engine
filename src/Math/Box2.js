import {
    JSON_stringify_fixedPrecision
} from './Common.js';
import {
    Vec2
} from './Vec2';
import {
    typeRegistry
} from './TypeRegistry.js';

class Box2 {
    constructor(p0 = undefined, p1 = undefined) {
        if (p0 instanceof Vec2) {
            this.p0 = p0;
        } else {
            this.p0 = new Vec2(Number.POSITIVE_INFINITY, Number.POSITIVE_INFINITY);
        }
        if (p1 instanceof Vec2) {
            this.p1 = p1;
        } else {
            this.p1 = new Vec2(Number.NEGATIVE_INFINITY, Number.NEGATIVE_INFINITY);
        }
    }


    set(p0, p1) {
        this.p0 = p0;
        this.p1 = p1;
    }

    reset() {
        this.p0.x = Number.POSITIVE_INFINITY;
        this.p1.x = Number.NEGATIVE_INFINITY;
        this.p0.y = Number.POSITIVE_INFINITY;
        this.p1.y = Number.NEGATIVE_INFINITY;
    }

    isValid() {
        return this.p0.x != Number.POSITIVE_INFINITY && this.p1.x != Number.NEGATIVE_INFINITY &&
            this.p0.y != Number.POSITIVE_INFINITY && this.p1.y != Number.NEGATIVE_INFINITY;
    }

    addPoint(point) {
        if (this.p0.x == Number.POSITIVE_INFINITY || point.x < this.p0.x)
            this.p0.x = point.x;
        if (this.p0.y == Number.POSITIVE_INFINITY || point.y < this.p0.y)
            this.p0.y = point.y;

        if (this.p1.y == Number.NEGATIVE_INFINITY || point.x > this.p1.x)
            this.p1.x = point.x;
        if (this.p1.y == Number.NEGATIVE_INFINITY || point.y > this.p1.y)
            this.p1.y = point.y;
    }

    size() {
        return this.p1.subtract(this.p0);
    }
    diagonal() {
        return this.p1.subtract(this.p0);
    }

    center() {
        let result = this.p1.subtract(this.p0);
        result.scaleInPlace(0.5);
        result.addInPlace(this.p0);
        return result;
    }


    //////////////////////////////////////////
    // Static Methods

    static create(...args) {
        return new Box2(...args);
    }

    //////////////////////////////////////////
    // Persistence

    toJSON() {
        return {
            "p0": this.p0.toJSON(),
            "p1": this.p1.toJSON()
        }
    }

    toString() {
        return JSON_stringify_fixedPrecision(this.toJSON())
    }
};

typeRegistry.registerType('Box2', Box2);

export {
    Box2
};
// export default Box2;