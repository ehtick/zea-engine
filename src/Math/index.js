
export {
    SInt32,
    UInt32,
    Float32,
    hashStr,
    JSON_stringify_fixedPrecision
} from './Common.js';
export * from './AttrValue.js';
export * from './Vec2.js';
export * from './Vec3.js';
export * from './Vec4.js';
export * from './Color.js';
export * from './Quat.js';
export * from './Ray.js';
export * from './Mat3.js';
export * from './Mat4.js';
export * from './Xfo.js';
export * from './Box2.js';
export * from './Box3.js';
export * from './Rect2.js';
export * from './Hammersley.js';

export * from './TypeRegistry.js';
