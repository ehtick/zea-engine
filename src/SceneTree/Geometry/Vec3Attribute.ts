import { Attribute } from './Attribute'
import { Vec3 } from '../../Math/Vec3'
import { Registry } from '../../Registry'

/**
 * Class representing an attribute.
 */
class Vec3Attribute extends Attribute {
  /**
   * Create a Vec3Attribute.
   */
  constructor(dataTypeName: string = 'Vec3') {
    super(dataTypeName, 3)
    this.normalized = false
  }

  /**
   * Returns the Vec3 value at the specified index.
   *
   * > Note: 'Ref' means that the value contains a reference to the data in the attribute.
   * > The components of the value can be changed causing the attributes data is changed.
   * > No need to call 'setValue'.
   *
   * @param index - The index value.
   * @returns Vec3 - The value at the specified index.
   */
  getValueRef(index: number): Vec3 {
    if (index >= this.data.length / this.stride)
      throw new Error('Invalid vertex index:' + index + '. Num Vertices:' + this.data.length / 3)

    const offset = index * this.stride
    const valueData = this.data.subarray(offset, offset + this.stride)
    const vec3 = new Vec3()
    Object.defineProperty(vec3, 'x', { get: () => valueData[0], set: (value) => (valueData[0] = value) })
    Object.defineProperty(vec3, 'y', { get: () => valueData[1], set: (value) => (valueData[1] = value) })
    Object.defineProperty(vec3, 'z', { get: () => valueData[2], set: (value) => (valueData[2] = value) })
    vec3.set = (x: number, y: number, z: number): void => {
      valueData[0] = x
      valueData[1] = y
      valueData[2] = z
    }
    return vec3
  }

  /**
   * Returns a copy of the Vec3 value at the specified index.
   *
   * @param index - The index value.
   * @return Vec3 - The value at the specified index.
   */
  getValue(index: number): Vec3 {
    if (index >= this.data.length / this.stride)
      throw new Error('Invalid vertex index:' + index + '. Num Vertices:' + this.data.length / 3)

    const offset = index * this.stride
    const valueData = this.data.slice(offset, offset + this.stride)
    return new Vec3(valueData[0], valueData[1], valueData[2])
  }

  /**
   * Sets Vec3 at the specified index.
   *
   * @param index - The index value.
   * @param value - The value param.
   */
  setValue(index: number, value: Vec3): void {
    if (index >= this.data.length / this.stride)
      throw new Error('Invalid vertex index:' + index + '. Num Vertices:' + this.data.length / 3)

    const offset = index * this.stride
    this.data.set(value.asArray(), offset)
  }

  /**
   * Gets the value of a corner vertex of a face.
   * > Note: 'Ref' means that the value contains a reference to the data in the attribute.
   * > The components of the value can be changed causing the attributes data is changed.
   * > No need to call 'setFaceVertexValue'.
   * @param face - The face index.
   * @param faceVertex - The index of vertex within the face. [0... num face vertices]
   * @return - The return value.
   */
  getFaceVertexValueRef(face: number, faceVertex: number): Vec3 {
    const valueData = this.getFaceVertexValueRef_array(face, faceVertex)
    const vec3 = new Vec3()
    Object.defineProperty(vec3, 'x', { get: () => valueData[0], set: (value) => (valueData[0] = value) })
    Object.defineProperty(vec3, 'y', { get: () => valueData[1], set: (value) => (valueData[1] = value) })
    Object.defineProperty(vec3, 'z', { get: () => valueData[2], set: (value) => (valueData[2] = value) })
    vec3.set = (x: number, y: number, z: number): void => {
      valueData[0] = x
      valueData[1] = y
      valueData[2] = z
    }
    return vec3
  }

  /**
   * Gets the value of a corner vertex of a face.
   * > Note: 'Ref' means that the value contains a reference to the data in the attribute.
   * > The components of the value can be changed causing the attributes data is changed.
   * > No need to call 'setFaceVertexValue'.
   * @param face - The face index.
   * @param faceVertex - The index of vertex within the face. [0... num face vertices]
   * @return - The return value.
   */
  getFaceVertexValue(face: number, faceVertex: number): Vec3 {
    const array = this.getFaceVertexValueRef_array(face, faceVertex)
    return new Vec3(array[0], array[1], array[2])
  }

  /**
   * Sets the value of a corner vertex of a face.
   * @param face - The face index.
   * @param faceVertex - The index of vertex within the face. [0... num face vertices]
   * @param value - The value value.
   */
  setFaceVertexValue(face: number, faceVertex: number, value: Vec3): void {
    this.setFaceVertexValue_array(face, faceVertex, value.asArray())
  }

  /**
   * The setSplitVertexValue method.
   * @param vertex - The vertex value.
   * @param face - The face index.
   * @param value - The value value.
   */
  setSplitVertexValue(vertex: number, face: number, value: Vec3): void {
    this.setSplitVertexValue_array(vertex, face, value.asArray())
  }
}

Registry.register('Vec3Attribute', Vec3Attribute)

export { Vec3Attribute }
