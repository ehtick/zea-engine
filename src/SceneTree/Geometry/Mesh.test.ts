import { Mesh } from './Mesh'
import { Vec3 } from '../../Math'
import { Vec3Attribute } from './Vec3Attribute'

describe('Mesh', () => {
  test('Check for default positions attribute.', () => {
    const mesh = new Mesh()
    expect(mesh.hasVertexAttribute('positions')).toBe(true)
    expect(mesh.getNumFaces()).toBe(0)
  })

  test('Check various face counts and offset', () => {
    const mesh = new Mesh()
    mesh.setFaceCounts([7, 5, 3])
    mesh.setFaceVertexIndices(3, [0, 1, 2])
    mesh.setFaceVertexIndices(9, [0, 1, 2, 3])
    mesh.setFaceVertexIndices(13, [0, 1, 2, 3, 4])
    expect(() => {
      mesh.setFaceVertexIndices(14, [0, 1, 2, 3, 4, 5])
    }).toThrow()

    expect(mesh.getFaceVertexCount(3)).toBe(3)
    expect(mesh.getFaceVertexCount(9)).toBe(4)
    expect(mesh.getFaceVertexCount(13)).toBe(5)
    expect(mesh.getFaceVertexIndices(3)).toEqual([0, 1, 2])
    expect(mesh.getFaceVertexIndices(9)).toEqual([0, 1, 2, 3])
    expect(mesh.getFaceVertexIndices(13)).toEqual([0, 1, 2, 3, 4])
  })

  test('Check adding a triangle', () => {
    const mesh = new Mesh()
    // The first face is a quad.
    mesh.addFace([0, 1, 2])
    expect(mesh.getNumFaces()).toBe(1)
    expect(mesh.getFaceVertexIndices(0)).toEqual([0, 1, 2])
  })

  test('Check adding a quad', () => {
    const mesh = new Mesh()
    // The first face is a quad.
    mesh.addFace([0, 1, 2, 3])
    expect(mesh.getNumFaces()).toBe(1)
    expect(mesh.getFaceVertexIndices(0)).toEqual([0, 1, 2, 3])
  })

  test('Check adding faces one at a time', () => {
    const mesh = new Mesh()

    mesh.addFace([0, 1, 2])
    mesh.addFace([0, 1, 2, 3])
    mesh.addFace([0, 1, 2, 3, 4])
    mesh.addFace([0, 1, 2])
    mesh.addFace([0, 1, 2, 3, 4, 5])

    expect(mesh.getFaceVertexCount(0)).toBe(3)
    expect(mesh.getFaceVertexCount(2)).toBe(4)
    expect(mesh.getFaceVertexCount(3)).toBe(5)
    expect(mesh.getFaceVertexCount(1)).toBe(3)
    expect(mesh.getFaceVertexIndices(0)).toEqual([0, 1, 2])
    expect(mesh.getFaceVertexIndices(2)).toEqual([0, 1, 2, 3])
    expect(mesh.getFaceVertexIndices(3)).toEqual([0, 1, 2, 3, 4])
    expect(mesh.getFaceVertexIndices(1)).toEqual([0, 1, 2])
    expect(mesh.getFaceVertexIndices(4)).toEqual([0, 1, 2, 3, 4, 5])
  })

  test('Check for setting up a single triangle', () => {
    const mesh = new Mesh()
    const numVertices = 3
    mesh.setNumVertices(numVertices)

    const positions = <Vec3Attribute>mesh.positions
    positions.getValueRef(0).set(1, 2, 3)
    positions.getValueRef(1).set(-1, -2, -3)
    positions.getValueRef(2).set(2, 1, -3)

    mesh.setFaceCounts([1])
    mesh.setFaceVertexIndices(0, [0, 1, 2])

    expect(mesh.getNumFaces()).toBe(1)
    expect(mesh.getFaceVertexCount(0)).toBe(3)
    expect(mesh.getFaceVertexIndices(0)).toEqual([0, 1, 2])
    expect(mesh.computeHardEdgesIndices()).toEqual(new Uint32Array([0, 1, 1, 2, 0, 2]))

    expect(mesh.positions.getCount()).toBe(numVertices)
  })

  test('Check for setting up a single quad', () => {
    const mesh = new Mesh()
    const numVertices = 4
    mesh.setNumVertices(numVertices)

    const positions = <Vec3Attribute>mesh.positions
    positions.getValueRef(0).set(0, -1, -1)
    positions.getValueRef(1).set(0, -1, 1)
    positions.getValueRef(2).set(0, 1, 1)
    positions.getValueRef(3).set(0, 1, -1)

    mesh.addVertexAttribute('normals', new Vec3Attribute())

    const normals = <Vec3Attribute>mesh.getVertexAttribute('normals')
    normals.getValueRef(0).set(-1, 0, 0)
    normals.getValueRef(1).set(-1, 0, 0)
    normals.getValueRef(2).set(-1, 0, 0)
    normals.getValueRef(3).set(-1, 0, 0)

    mesh.setFaceCounts([0, 1])
    mesh.setFaceVertexIndices(0, [0, 1, 2, 3])
    expect(mesh.computeHardEdgesIndices()).toEqual(new Uint32Array([0, 1, 1, 2, 2, 3, 0, 3]))

    expect(mesh.getFaceVertexIndices(0)).toEqual([0, 1, 2, 3])
  })

  test('Check for setting up 2 neighboring quads', () => {
    const mesh = new Mesh()
    const numVertices = 6

    mesh.addVertexAttribute('normals', new Vec3Attribute())
    mesh.setNumVertices(numVertices)

    const positions = <Vec3Attribute>mesh.positions
    positions.getValueRef(0).set(0, -1, -1)
    positions.getValueRef(1).set(0, -1, 1)
    positions.getValueRef(2).set(0, 1, 1)
    positions.getValueRef(3).set(0, 1, -1)
    positions.getValueRef(4).set(1, 1, 1)
    positions.getValueRef(5).set(1, 1, -1)

    mesh.setFaceCounts([0, 2])
    mesh.setFaceVertexIndices(0, [0, 1, 2, 3])
    mesh.setFaceVertexIndices(1, [3, 2, 4, 5])

    const normals = <Vec3Attribute>mesh.getVertexAttribute('normals')

    // The first face has normals along the -X axis
    normals.setFaceVertexValue(0, 0, new Vec3(-1, 0, 0))
    normals.setFaceVertexValue(0, 1, new Vec3(-1, 0, 0))
    normals.setFaceVertexValue(0, 2, new Vec3(-1, 0, 0))
    normals.setFaceVertexValue(0, 3, new Vec3(-1, 0, 0))

    // The second face has normals along the Y axis
    normals.setFaceVertexValue(1, 0, new Vec3(0, 1, 0))
    normals.setFaceVertexValue(1, 1, new Vec3(0, 1, 0))
    normals.setFaceVertexValue(1, 2, new Vec3(0, 1, 0))
    normals.setFaceVertexValue(1, 3, new Vec3(0, 1, 0))

    expect(mesh.getFaceVertexIndices(0)).toEqual([0, 1, 2, 3])
    expect(mesh.getFaceVertexIndices(1)).toEqual([3, 2, 4, 5])
  })

  test('Check computing normals on 2 neighboring quads', () => {
    const mesh = new Mesh()
    const numVertices = 6
    mesh.setNumVertices(numVertices)

    const positions = <Vec3Attribute>mesh.positions
    positions.getValueRef(0).set(0, -1, -1)
    positions.getValueRef(1).set(0, -1, 1)
    positions.getValueRef(2).set(0, 1, 1)
    positions.getValueRef(3).set(0, 1, -1)
    positions.getValueRef(4).set(1, 1, 1)
    positions.getValueRef(5).set(1, 1, -1)

    mesh.setFaceCounts([0, 2, 0])
    mesh.setFaceVertexIndices(0, [0, 1, 2, 3])
    mesh.setFaceVertexIndices(1, [3, 2, 4, 5])

    mesh.computeVertexNormals()

    // Now check that the 2 faces have normals
    const faceNormals = <Vec3Attribute>mesh.getFaceAttribute('normals')
    expect(faceNormals.getValueRef(0)).toEqual(new Vec3(-1, 0, 0))
    expect(faceNormals.getValueRef(1)).toEqual(new Vec3(0, 1, 0))

    // Now we check that the 2 normal values for the ve
    // vertex id 2 is used by Face 0 at faceVertex 2, and Face 1 and faceVertex index 1
    // vertex id 3 is used by Face 0 at faceVertex 3, and Face 1 and faceVertex index 0
    const normals = <Vec3Attribute>mesh.getVertexAttribute('normals')
    expect(normals.getFaceVertexValueRef(0, 2)).toEqual(new Vec3(-1, 0, 0))
    expect(normals.getFaceVertexValueRef(1, 1)).toEqual(new Vec3(0, 1, 0))
    expect(normals.getFaceVertexValueRef(0, 3)).toEqual(new Vec3(-1, 0, 0))
    expect(normals.getFaceVertexValueRef(1, 0)).toEqual(new Vec3(0, 1, 0))

    expect(JSON.stringify(mesh.toJSON())).toMatchSnapshot()
  })

  test('Check computing edge indices', () => {
    const mesh = new Mesh()
    const numVertices = 6
    mesh.setNumVertices(numVertices)

    const positions = <Vec3Attribute>mesh.positions
    positions.getValueRef(0).set(0, -1, -1)
    positions.getValueRef(1).set(0, -1, 1)
    positions.getValueRef(2).set(0, 1, 1)
    positions.getValueRef(3).set(0, 1, -1)
    positions.getValueRef(4).set(0, 2, 1)
    positions.getValueRef(5).set(0, 2, -1)

    mesh.addFace([0, 1, 2, 3])
    mesh.addFace([3, 2, 4, 5])

    expect(mesh.computeHardEdgesIndices()).toEqual(new Uint32Array([0, 1, 1, 2, 0, 3, 2, 4, 4, 5, 3, 5]))
  })

  test('Check computing edge indices (hard edge in the middle)', () => {
    const mesh = new Mesh()
    const numVertices = 6
    mesh.setNumVertices(numVertices)

    const positions = <Vec3Attribute>mesh.positions
    positions.getValueRef(0).set(0, -1, -1)
    positions.getValueRef(1).set(0, -1, 1)
    positions.getValueRef(2).set(0, 1, 1)
    positions.getValueRef(3).set(0, 1, -1)
    positions.getValueRef(4).set(1, 1, 1)
    positions.getValueRef(5).set(1, 1, -1)

    mesh.addFace([0, 1, 2, 3])
    mesh.addFace([3, 2, 4, 5])

    expect(mesh.computeHardEdgesIndices()).toEqual(new Uint32Array([0, 1, 1, 2, 2, 3, 0, 3, 2, 4, 4, 5, 3, 5]))
  })

  test('Check resizing bigger the face indices.', () => {
    const mesh = new Mesh()
    const numVertices = 4
    mesh.setNumVertices(numVertices)

    // one triangle and one quad.
    mesh.setFaceCounts([1, 1])
    mesh.setFaceVertexIndices(0, [0, 1, 2])
    mesh.setFaceVertexIndices(1, [0, 1, 2, 3])

    // Now we add another triangle
    mesh.setFaceCounts([2, 1])
    mesh.setFaceVertexIndices(1, [2, 1, 3])

    expect(mesh.getFaceVertexIndices(0)).toEqual([0, 1, 2])
    expect(mesh.getFaceVertexIndices(1)).toEqual([2, 1, 3])
    expect(mesh.getFaceVertexIndices(2)).toEqual([0, 1, 2, 3])
  })

  test.skip('Check resizing smaller the face indices.', () => {
    const mesh = new Mesh()
    const numVertices = 4
    mesh.setNumVertices(numVertices)

    // two triangleas and one quad.
    mesh.setFaceCounts([2, 1])
    mesh.setFaceVertexIndices(0, [0, 1, 2])
    mesh.setFaceVertexIndices(1, [2, 1, 3])
    mesh.setFaceVertexIndices(1, [2, 1, 3])

    // remove a triangle. Now one triangle and one quad.
    mesh.setFaceCounts([1, 1])

    expect(mesh.getFaceVertexIndices(0)).toEqual([0, 1, 2])
    expect(mesh.getFaceVertexIndices(1)).toEqual([0, 1, 2, 3])
  })

  test('Check generated buffers', () => {
    const mesh = new Mesh()
    const numVertices = 3
    mesh.setNumVertices(numVertices)

    const positions = <Vec3Attribute>mesh.positions
    positions.getValueRef(0).set(1, 2, 3)
    positions.getValueRef(1).set(-1, -2, -3)
    positions.getValueRef(2).set(2, 1, -3)

    mesh.setFaceCounts([1])
    mesh.setFaceVertexIndices(0, [0, 1, 2])

    expect(mesh.genBuffers()).toEqual({
      attrBuffers: {
        positions: {
          count: 3,
          dataType: 'Vec3',
          dimension: 3,
          normalized: false,
          values: new Float32Array([1, 2, 3, -1, -2, -3, 2, 1, -3]),
        },
      },
      indices: new Uint8Array([0, 1, 2]),
      numRenderVerts: 3,
      numVertices: 3,
    })
  })

  test('saves to JSON (serialization).', () => {
    const mesh = new Mesh()
    const numVertices = 3
    mesh.setNumVertices(numVertices)

    const positions = <Vec3Attribute>mesh.positions
    positions.getValueRef(0).set(1, 2, 3)
    positions.getValueRef(1).set(-1, -2, -3)
    positions.getValueRef(2).set(2, 1, -3)

    mesh.setFaceCounts([1])
    mesh.setFaceVertexIndices(0, [0, 1, 2])

    expect(JSON.stringify(mesh.toJSON())).toMatchSnapshot()
  })

  test('loads from JSON (serialization).', () => {
    const mesh = new Mesh()
    const input: Record<string, any> = {
      type: 'Mesh',
      numVertices: 3,
      vertexAttributes: {
        positions: {
          data: [1, 2, 3, -1, -2, -3, 2, 1, -3],
          dataType: 'Vec3',
          defaultValue: 0,
          length: 3,
          splits: {},
          splitValues: [],
        },
      },
      faceCounts: [1],
      faceVertexIndices: [0, 1, 2],
    }
    mesh.fromJSON(input)

    expect(mesh.positions.getCount()).toBe(3)
    expect(mesh.getFaceCounts()).toEqual([1])
    expect(mesh.getFaceVertexIndices(0)).toEqual([0, 1, 2])
  })
})
