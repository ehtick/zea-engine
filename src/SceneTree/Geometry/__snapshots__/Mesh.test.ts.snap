// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mesh Check computing normals on 2 neighboring quads 1`] = `"{\\"type\\":\\"Mesh\\",\\"numVertices\\":6,\\"vertexAttributes\\":{\\"positions\\":{\\"data\\":{\\"0\\":0,\\"1\\":48128,\\"2\\":48128,\\"3\\":0,\\"4\\":48128,\\"5\\":15360,\\"6\\":0,\\"7\\":15360,\\"8\\":15360,\\"9\\":0,\\"10\\":15360,\\"11\\":48128,\\"12\\":15360,\\"13\\":15360,\\"14\\":15360,\\"15\\":15360,\\"16\\":15360,\\"17\\":48128},\\"dataType\\":\\"Vec3f16\\",\\"length\\":6},\\"normals\\":{\\"data\\":{\\"0\\":-127,\\"1\\":0,\\"2\\":0,\\"3\\":-127,\\"4\\":0,\\"5\\":0,\\"6\\":-127,\\"7\\":0,\\"8\\":0,\\"9\\":0,\\"10\\":127,\\"11\\":0,\\"12\\":0,\\"13\\":127,\\"14\\":0,\\"15\\":0,\\"16\\":127,\\"17\\":0},\\"dataType\\":\\"Vec3f8\\",\\"length\\":6}},\\"faceCounts\\":[0,2,0],\\"faceVertexIndices\\":[0,1,2,3,3,2,4,5]}"`;

exports[`Mesh saves to JSON (serialization). 1`] = `"{\\"type\\":\\"Mesh\\",\\"numVertices\\":3,\\"vertexAttributes\\":{\\"positions\\":{\\"data\\":{\\"0\\":15360,\\"1\\":16384,\\"2\\":16896,\\"3\\":48128,\\"4\\":49152,\\"5\\":49664,\\"6\\":16384,\\"7\\":15360,\\"8\\":49664},\\"dataType\\":\\"Vec3f16\\",\\"length\\":3}},\\"faceCounts\\":[1],\\"faceVertexIndices\\":[0,1,2]}"`;
